### ISSUE TITLE
API to support generate random number?

### ISSUE THREAD

**Issue by xiongyu0523** [raspberrypi/pico-sdk#569]
is there a plan to add hardware random number generation?

**Comment by lurch** [comment]
Are you looking for "2.17.5. Random Number Generator" in https://datasheets.raspberrypi.org/rp2040/rp2040-datasheet.pdf ?
Or are you asking for a "nice API" to wrap that?

**Comment by Wren6991** [comment]
You can read the ring oscillator random bit with something like

```c
#include "hardware/structs.rosc.h"

bool get_random_bit() {
    return rosc_hw->randombit;
}
```

This is just an unconditioned ring oscillator sample. It's not well characterised, somewhat biased, output is somewhat periodic when sampled rapidly, etc. We haven't wrapped it up in a simplified API because that might give an impression of randomness guarantees that we haven't done the characterisation to back up. It's very much "some assembly required".

What functionality are you looking for exactly?

**Comment by xiongyu0523** [comment]
Thank you very much for your example and forum discussions. I will use the rand number for the SSL library port. I can understand the hardware characteristic is not defined yet, but providing such a API is still useful and can be a seed feed to srand

**Comment by makeworld-the-better-one** [comment]
This is fixed thanks to #4213. You can generate randomness on the RP2040 using `os.urandom` .

**Comment by lurch** [comment]
@makeworld-the-better-one It sounds like you might be mixing up the RP2040 C SDK with the RP2040 MicroPython port?

**Comment by makeworld-the-better-one** [comment]
Yep sorry, my bad

**Comment by kilograham** [comment]
This seems like a useful addition, as the random bit is not sufficiently random if polled at high frequency

**Comment by cameronelliott** [comment]
> Thank you very much for your example and forum discussions. I will use the rand number for the SSL library port. I can understand the hardware characteristic is not defined yet, but providing such a API is still useful and can be a seed feed to srand

You did see the note in the forums? "The datsheet does say that the randomness is not suitable for cryptography, since it can be compromised, and thus is only suitable for less critical applications."

**Comment by kilograham** [comment]
fixed by #1111

**Comment by kilograham** [comment]
merged into develop
